nilearn/glm/tests/test_utils.py

```diff
@@ -31,12 +37,101 @@ def test_full_rank():
+    tval = np.random.standard_t(1e10, size=10)
+    # Estimate the p-values using the Survival Function (SF)
+    pval = sps.t.sf(tval, 1e10)
+    # Estimate the p-values using the Cumulative Distribution Function (CDF)
+    cdfval = sps.t.cdf(tval, 1e10)
+    # Set a minimum threshold for p-values to avoid infinite z-scores
+    pval = np.array(np.minimum(np.maximum(pval, 1.e-300), 1. - 1.e-16))
+    cdfval = np.array(np.minimum(np.maximum(cdfval, 1.e-300), 1. - 1.e-16))
+    # Compute z-score from the p-value estimated with the SF
+    zval_sf = norm.isf(pval)
+    # Compute z-score from the p-value estimated with the CDF
+    zval_cdf = norm.ppf(cdfval)
+    # Create the final array of z-scores, ...
+    zval = np.zeros(pval.size)
+    # ... in which z-scores < 0 estimated w/ SF are replaced by z-scores < 0
+    # estimated w/ CDF
+    zval[np.atleast_1d(zval_sf < 0)] = zval_cdf[zval_sf < 0]
+    # ... and z-scores >=0 estimated from SF are kept.
+    zval[np.atleast_1d(zval_sf >= 0)] = zval_sf[zval_sf >= 0]
+    # Test 'z_score' function in 'nilearn/glm/contrasts.py'
+    assert_array_almost_equal(z_score(pval, one_minus_pvalue=cdfval), zval)
+    # ################# Check z-scores computed from F-values #################
+    # Randomly draw samples from the F distribution
+    fval = np.random.f(1e10, 42, size=10)
+    # Estimate the p-values using the Survival Function (SF)
+    p_val = sps.f.sf(fval, 42, 1e10)
+    # Estimate the p-values using the Cumulative Distribution Function (CDF)
+    cdf_val = sps.f.cdf(fval, 42, 1e10)
+    # Set a minimum threshold for p-values to avoid infinite z-scores
+    p_val = np.array(np.minimum(np.maximum(p_val, 1.e-300), 1. - 1.e-16))
+    cdf_val = np.array(np.minimum(np.maximum(cdf_val, 1.e-300), 1. - 1.e-16))
+    # Compute z-score from the p-value estimated with the SF
+    z_val_sf = norm.isf(p_val)
+    # Compute z-score from the p-value estimated with the CDF
+    z_val_cdf = norm.ppf(cdf_val)
+    # Create the final array of z-scores, ...
+    z_val = np.zeros(p_val.size)
+    # ... in which z-scores < 0 estimated w/ SF are replaced by z-scores < 0
+    # estimated w/ CDF
+    z_val[np.atleast_1d(z_val_sf < 0)] = z_val_cdf[z_val_sf < 0]
+    # ... and z-scores >=0 estimated from SF are kept.
+    z_val[np.atleast_1d(z_val_sf >= 0)] = z_val_sf[z_val_sf >= 0]
+    # Test 'z_score' function in 'nilearn/glm/contrasts.py'
+    assert_array_almost_equal(z_score(p_val, one_minus_pvalue=cdf_val), z_val)
+    # ##################### Check the numerical precision #####################
+    for t in [33.75, -8.3]:
+        p = sps.t.sf(t, 1e10)
+        cdf = sps.t.cdf(t, 1e10)
+        z_sf = norm.isf(p)
+        z_cdf = norm.ppf(cdf)
+        if p <= .5:
+            z = z_sf
+        else:
+            z = z_cdf
+        assert_array_almost_equal(z_score(p, one_minus_pvalue=cdf), z)
+
+
+def test_z_score_opposite_contrast():
+    fmri, mask = generate_fake_fmri(shape=(50, 20, 50), length=96)
```

Comment:
could you create a new random number generator and pass that to `generate_fake_fmri` ? the default generator used by generate_fake_fmri is only initialized once, so the result may depend on the order in which tests are executed. I'll open a separate issue about that

Reply:
I assigned the arg `rand_gen=np.random.RandomState(42)` in `generate_fake_fmri` function? It is by default set to 0. Was that what you asked for?

Reply:
yes, that works! by default it is not set to 0, but to a mutable object `RandomState(0)` which is initialized once and then used/ modified by several test functions in unpredictable order, see #2631 .